### PR TITLE
Extract object parsing to separate class

### PR DESCRIPTION
**Pull request by vjik**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ✔️
| Breaks BC?    | ✔️
| Fixed issues  | #365

### CODE REVIEW

File: src/DataSet/ObjectDataSet.php [hunk 1]
```diff
@@ -26,32 +19,22 @@ final class ObjectDataSet implements RulesProviderInterface, DataSetInterface
...
-    private string|null $cacheKey = null;
+    private ObjectParser $parser;
 
     public function __construct(
         private object $object,
         private int $propertyVisibility = ReflectionProperty::IS_PRIVATE |
         ReflectionProperty::IS_PROTECTED |
         ReflectionProperty::IS_PUBLIC,
-        private bool $useCache = true
+        bool $useCache = true
```
Comment (arogachev) [inline src/DataSet/ObjectDataSet.php:29]: Could you add test with `$useCache = false`?

File: src/Helper/ObjectParser.php [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+        if ($this->hasCacheItem('reflectionProperties')) {
+            /** @var array<string, ReflectionProperty> */
+            return $this->getCacheItem('reflectionProperties');
+        }
+
+        $reflection = new ReflectionObject($this->object);
+        $reflectionProperties = [];
+
+        foreach ($reflection->getProperties($this->propertyVisibility) as $property) {
+            if ($this->skipStaticProperties && $property->isStatic()) {
```
Comment (arogachev) [inline src/Helper/ObjectParser.php:110]: Could you add test to check skipping of static properties?

File: src/DataSet/ObjectDataSet.php [hunk 3]
```diff
@@ -26,32 +19,22 @@ final class ObjectDataSet implements RulesProviderInterface, DataSetInterface
...
-            'reflectionAttributes' => 'array',
-        ],
-    ])]
-    private static array $cache = [];
-    private string|null $cacheKey = null;
+    private ObjectParser $parser;
 
     public function __construct(
         private object $object,
         private int $propertyVisibility = ReflectionProperty::IS_PRIVATE |
```
Comment (arogachev) [inline src/DataSet/ObjectDataSet.php:26]:
```suggestion
        int $propertyVisibility = ReflectionProperty::IS_PRIVATE |
```

Comment (arogachev) [reply src/DataSet/ObjectDataSet.php:26]: > Property is only written but never read

File: src/Helper/ObjectParser.php [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+        private object $object,
+        private int $propertyVisibility = ReflectionProperty::IS_PRIVATE |
+        ReflectionProperty::IS_PROTECTED |
+        ReflectionProperty::IS_PUBLIC,
+        private bool $skipStaticProperties = false,
+        bool $useCache = true
+    ) {
+        $this->cacheKey = $useCache
+            ? $this->object::class . '_' . $this->propertyVisibility . '_' . $this->skipStaticProperties
+            : null;
```
Comment (arogachev) [inline src/Helper/ObjectParser.php:41]: Could you add test with `null` cache key? Related - https://github.com/yiisoft/validator/pull/373/files#r1031517299.

File: src/Helper/ObjectParser.php [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+use function array_key_exists;
+
+final class ObjectParser
+{
+    /**
+     * @psalm-var array<string, array<string, array>>
+     */
+    #[ArrayShape([
+        [
+            'rules' => 'array',
```
Comment (arogachev) [inline src/Helper/ObjectParser.php:24]: It was an iterable before. How we can guarantee that it's always an array?

File: src/Helper/ObjectParser.php [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+use ReflectionProperty;
+use Yiisoft\Validator\AfterInitAttributeEventInterface;
+use Yiisoft\Validator\RuleInterface;
+
+use function array_key_exists;
+
+final class ObjectParser
+{
+    /**
+     * @psalm-var array<string, array<string, array>>
```
Comment (arogachev) [inline src/Helper/ObjectParser.php:20]: `psalm-` prefix can be omitted.

File: src/Helper/ObjectParser.php [hunk 7]
```diff
@@ -0,0 +1,167 @@
+<?php
```
Comment (arogachev) [inline src/Helper/ObjectParser.php:1]: Maybe this should be a part of other package, what do you think? Or validator is fine?

**Review by arogachev — CHANGES_REQUESTED** [review]
:+1:

File: src/Helper/ObjectParser.php (hunk 5 above)
Comment (vjik) [reply src/Helper/ObjectParser.php:24]: [Here](https://github.com/yiisoft/validator/blob/1304c80c9f2343bde15cd3db7f6b732708c4ba54/src/Helper/ObjectParser.php#L69) rules always array.

File: src/Helper/ObjectParser.php (hunk 7 above)
Comment (vjik) [reply src/Helper/ObjectParser.php:1]: It's fine, because parser do validator-specific parsing (collect rules and run rules events),

File: src/Helper/ObjectParser.php (hunk 6 above)
Comment (vjik) [reply src/Helper/ObjectParser.php]: done

File: src/Helper/ObjectParser.php (hunk 4 above)
Comment (vjik) [reply src/Helper/ObjectParser.php:41]: done

File: src/DataSet/ObjectDataSet.php (hunk 3 above)
Comment (vjik) [reply src/DataSet/ObjectDataSet.php]: done

File: src/Helper/ObjectParser.php (hunk 2 above)
Comment (vjik) [reply src/Helper/ObjectParser.php:110]: done

File: tests/DataSet/ObjectDataSetTest.php [hunk 8]
```diff
@@ -401,6 +402,20 @@ public function testCaching(array $objectDataSets, int $expectedRulesCallsCount,
         $this->assertSame($expectedDataCallsCount, ObjectWithCallsCount::$getDataCallsCount);
     }
 
+    public function testWithoutCache(): void
+    {
+        $object1 = new stdClass();
```
Comment (arogachev) [inline tests/DataSet/ObjectDataSetTest.php:407]: Maybe use data provider?

File: tests/Helper/ObjectParserTest.php [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+final class ObjectParserTest extends TestCase
+{
+    public function dataSkipStaticProperties(): array
+    {
+        return [
+            [
+                ['a' => 4, 'b' => 2],
+                new class () {
+                    public int $a = 4;
+                    public static int $b = 2;
```
Comment (arogachev) [inline tests/Helper/ObjectParserTest.php:19]: Let's add non-static "$c" property here to make sure the loop doesn't stop and continues.

File: tests/Helper/ObjectParserTest.php [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+                },
+                false,
+            ],
+            [
+                ['a' => 4],
+                new class () {
+                    public int $a = 4;
+                    public static int $b = 2;
```
Comment (arogachev) [inline tests/Helper/ObjectParserTest.php:27]: Let's add non-static "$c" property here to make sure the loop doesn't stop and continues. Can be added to the first data set too.

File: src/Helper/ObjectParser.php [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+    private static array $cache = [];
+    private string|null $cacheKey;
+
+    public function __construct(
+        private object $object,
+        private int $propertyVisibility = ReflectionProperty::IS_PRIVATE |
+        ReflectionProperty::IS_PROTECTED |
+        ReflectionProperty::IS_PUBLIC,
+        private bool $skipStaticProperties = false,
+        bool $useCache = true
```
Comment (arogachev) [inline src/Helper/ObjectParser.php:37]: 1 escaped mutant is left (changing default value from `true` to `false`). Need to adjust / actualize`ObjectDataSetTest::testCaching()` because the technique used there no longer works.

**Review by arogachev — CHANGES_REQUESTED** [review]
Some tests can be improved or need to be adjusted / actualized.

File: tests/DataSet/ObjectDataSetTest.php (hunk 8 above)
Comment (vjik) [reply tests/DataSet/ObjectDataSetTest.php:407]: It's cache testing. This is should work without dataProvide in one test.

File: tests/Helper/ObjectParserTest.php (hunk 10 above)
Comment (vjik) [reply tests/Helper/ObjectParserTest.php:27]: done

File: src/Helper/ObjectParser.php (hunk 11 above)
Comment (vjik) [reply src/Helper/ObjectParser.php:37]: done

File: tests/Support/Data/RuleWithCalls.php [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+
+namespace Yiisoft\Validator\Tests\Support\Data;
+
+use Attribute;
+use RuntimeException;
+use Yiisoft\Validator\AfterInitAttributeEventInterface;
+use Yiisoft\Validator\RuleInterface;
+
+#[Attribute(Attribute::TARGET_PROPERTY)]
+final class RuleWithCalls implements RuleInterface, AfterInitAttributeEventInterface
```
Comment (arogachev) [inline tests/Support/Data/RuleWithCalls.php:13]:
```suggestion
final class RuleWithCallsCount implements RuleInterface, AfterInitAttributeEventInterface
```

File: tests/Support/Data/RuleWithCalls.php [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+{
+    public static int $countCalls = 0;
```
Comment (arogachev) [inline tests/Support/Data/RuleWithCalls.php:15]:
```suggestion
    public static int $afterInitAttributeCallsCount = 0;
```

File: tests/Helper/ObjectParserTest.php [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+    {
+        $object = new class () {
+            public int $a = 4;
+            public static int $b = 2;
+        };
+
+        $parser = new ObjectParser($object);
+
+        $this->assertSame(['a' => 4, 'b' => 2], $parser->getData());
+    }
```
Comment (arogachev) [inline tests/Helper/ObjectParserTest.php:55]:
```suggestion
    public function dataGetData(): array
    {
        $object = new class () {
            public int $a = 4;
            public static int $b = 2;
            public string $c = 'hello';
        };

        return [
            [
                ['a' => 4, 'b' => 2, 'c' => 'hello'],
                $object,
                null,
            ],
            [
                ['a' => 4, 'b' => 2, 'c' => 'hello'],
                $object,
                false,
            ],
            [
                ['a' => 4, 'c' => 'hello'],
                $object,
                true,
            ],
        ];
    }

    /**
     * @dataProvider dataGetData
     */
    public function testGetData(array $expectedData, object $object, bool|null $skipStaticProperties): void
    {
        $arguments = [$object];
        if ($skipStaticProperties !== null) {
            $arguments['skipStaticProperties'] = $skipStaticProperties;
        }

        $parser = new ObjectParser(...$arguments);
        $this->assertSame($expectedData, $parser->getData());
    }
```

File: tests/Support/Data/RuleWithCalls.php [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+
+    public function afterInitAttribute(object $object): void
+    {
+        self::$countCalls++;
+    }
+
+    public function getName(): string
+    {
+        return 'rule-with-calls';
```
Comment (arogachev) [inline tests/Support/Data/RuleWithCalls.php:24]:
```suggestion
        return 'rule-with-calls-count';
```

**Review by arogachev — CHANGES_REQUESTED** [review]
Added 4 suggestions.

File: tests/Helper/ObjectParserTest.php (hunk 14 above)
Comment (vjik) [reply tests/Helper/ObjectParserTest.php:55]: Additional conditions — more difficult code. Suggest keep as is.

File: tests/Support/Data/RuleWithCalls.php (hunk 12 above)
Comment (vjik) [reply tests/Support/Data/RuleWithCalls.php]: done

File: tests/Support/Data/RuleWithCalls.php (hunk 13 above)
Comment (vjik) [reply tests/Support/Data/RuleWithCalls.php]: done

File: tests/Support/Data/RuleWithCalls.php (hunk 15 above)
Comment (vjik) [reply tests/Support/Data/RuleWithCalls.php]: done

**Review by arogachev — APPROVED** [review]
